### PR TITLE
Making the Uploads API public

### PR DESCRIPTION
**Pull request by kinelski**
Exposing the Uploads API as `Experimental` protocol methods, and adding tests to them.

### CODE REVIEW

File: .dotnet/tests/Utility/MultipartFormDataBinaryContent.cs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+using System.ClientModel;
+using System.IO;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace OpenAI.Tests;
+
+internal class MultipartFormDataBinaryContent : BinaryContent
```
Comment (kinelski) [inline .dotnet/tests/Utility/MultipartFormDataBinaryContent.cs:11]: Mostly copy-pasted from our source project.

File: .dotnet/tests/Files/FileTests.Uploads.cs [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        ClientResult result = IsAsync
+            ? await fileClient.CompleteUploadAsync(createdUploadDetails.Id, content)
+            : fileClient.CompleteUpload(createdUploadDetails.Id, content);
+
+        BinaryData response = result.GetRawResponse().Content;
+        using JsonDocument jsonDocument = JsonDocument.Parse(response);
+        JsonElement fileElement = jsonDocument.RootElement.GetProperty("file");
+        string fileId = fileElement.GetProperty("id").GetString();
+
+        await fileClient.DeleteFileAsync(fileId);
```
Comment (joseharriaga) [inline .dotnet/tests/Files/FileTests.Uploads.cs:106]: Before deleting the file, would it be possible to use the FileClient's DownloadFile method to validate that the content is equal to the first part and second part in the right order?

**Review by joseharriaga — APPROVED** [review]

**Comment by joseharriaga** [comment]
@trrwilson: I approved this PR, but you looked at Uploads longer than me, so it would be great if you could review it too. 😊

File: .dotnet/src/Custom/Files/FileClient.Protocol.cs [hunk 3]
```diff
@@ -209,4 +210,52 @@ public virtual ClientResult DownloadFile(string fileId, RequestOptions options)
         using PipelineMessage message = CreateDownloadFileRequest(fileId, options);
         return ClientResult.FromResponse(_pipeline.ProcessMessage(message, options));
     }
+
+    [Experimental("OPENAI001")]
+    public virtual async Task<ClientResult> CreateUploadAsync(BinaryContent content, RequestOptions options = null)
```
Comment (joseharriaga) [inline .dotnet/src/Custom/Files/FileClient.Protocol.cs:215]: Could you add some simple doc comments for these methods?

File: .dotnet/tests/Files/FileTests.Uploads.cs (hunk 2 above)
Comment (kinelski) [reply .dotnet/tests/Files/FileTests.Uploads.cs:106]: I gave it a try, but I'm facing some issues when downloading the file.
- If I set the purpose to "vision", upload creation fails with a 500 error (not on us).
- If I set the purpose to anything else, we get the following error when trying to download the file:
```
Not allowed to download files of purpose: <purpose>
```

I'm not sure if that's a bug or a limitation of our subscription, but it doesn't seem to be our fault.

File: .dotnet/src/Custom/Files/FileClient.Protocol.cs (hunk 3 above)
Comment (kinelski) [reply .dotnet/src/Custom/Files/FileClient.Protocol.cs:224]: Done!

File: .dotnet/tests/Files/FileTests.Uploads.cs (hunk 2 above)
Comment (joseharriaga) [reply .dotnet/tests/Files/FileTests.Uploads.cs:106]: Ah, I think this is a limitation of the API. Alright, no worries. Thank you for looking into it!

**Review by trrwilson — APPROVED** [review]

**Review by joseharriaga — APPROVED** [review]

**Pull request merged**